File: CTY_2026_JASA--empapp_tables.R
################################################################################################
# Estimation and Inference in Boundary Discontinuity Designs: Location-Based Methods
# Empirical Application: SPP data
# Tables
################################################################################################

rm(list = ls(all = TRUE))

# This script converts the CSV files produced by CTY_2026_JASA--empapp.R into
# manuscript-ready LaTeX tabular fragments. The full table keeps every
# evaluation point; the small table keeps selected evaluation points plus the
# WBATE and LBATE rows.

get_script_dir <- function() {
  file_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
  if (length(file_arg) > 0) {
    return(normalizePath(dirname(sub("^--file=", "", file_arg[1])), winslash = "/"))
  }
  normalizePath(getwd(), winslash = "/")
}

setwd(get_script_dir())

output_dir <- "output"
tables_dir <- "tables"
small_rows <- c(1, 5, 10, 15, 20, 25, 30, 35, 40)
aggregate_labels <- c("WBATE", "LBATE")

dir.create(tables_dir, showWarnings = FALSE)

# One row per CSV output that should become a pair of LaTeX tables. File names
# are intentionally tied to the source CSV stems.
table_specs <- data.frame(
  id = c(
    "empapp_main_fuzzy",
    "empapp_main_itt",
    "empapp_main_fs",
    "empapp_main_itt0",
    "empapp_covbal_itt",
    "empapp_main_percent"
  ),
  kind = c("effect", "effect", "effect", "effect", "effect", "percent"),
  small = c(TRUE, TRUE, TRUE, TRUE, TRUE, FALSE),
  stringsAsFactors = FALSE
)

bw_specs <- data.frame(
  bwparam = c("main", "itt"),
  suffix = c("bwmain", "bwitt"),
  stringsAsFactors = FALSE
)

################################## Input #######################################

clean_tables_dir <- function() {
  # Generated tables should be reproducible from output/*.csv, so remove stale
  # empirical fragments before writing the current set.
  old_files <- list.files(tables_dir, pattern = "^empapp_.*\\.tex$", full.names = TRUE)
  if (length(old_files) > 0) unlink(old_files)
}

read_result_table <- function(id, kind, suffix) {
  path <- file.path(output_dir, paste0(id, "_", suffix, ".csv"))
  if (!file.exists(path)) {
    stop(sprintf("Missing output file: %s. Run CTY_2026_JASA--empapp.R first.", path), call. = FALSE)
  }

  tab <- read.csv(path, check.names = FALSE)
  required <- c("row", "h01", "h02", "h11", "h12", "N.Co", "N.Tr")
  if (kind == "effect") {
    required <- c(required, "estimate.p", "p.value", "ci.lower", "ci.upper")
  } else if (kind == "percent") {
    required <- c(required, "itt.percent", "fuzzy.percent")
  } else {
    stop(sprintf("Unknown table kind: %s", kind), call. = FALSE)
  }
  missing <- setdiff(required, names(tab))
  if (length(missing) > 0) {
    stop(sprintf("%s is missing columns: %s", path, paste(missing, collapse = ", ")), call. = FALSE)
  }

  tab$row <- as.character(tab$row)
  tab
}

################################ Formatting ####################################

is_aggregate_row <- function(row_label) row_label %in% aggregate_labels

fmt_num <- function(x, digits = 3) {
  if (is.na(x) || !is.finite(x)) return("")
  sprintf(paste0("%.", digits, "f"), x)
}

fmt_pvalue <- function(x) {
  if (is.na(x) || !is.finite(x)) return("")
  if (x < 0.001) return("0.000")
  sprintf("%.3f", x)
}

fmt_interval <- function(lower, upper) {
  if (is.na(lower) || is.na(upper) || !is.finite(lower) || !is.finite(upper)) return("")
  sprintf("(%s, %s)", fmt_num(lower), fmt_num(upper))
}

fmt_row_label <- function(row_id) {
  if (row_id == "WBATE") return("$\\mathtt{WBATE}$")
  if (row_id == "LBATE") return("$\\mathtt{LBATE}$")
  sprintf("$\\bb_{%s}$", row_id)
}

bandwidth_values <- function(tab, i) {
  if (is_aggregate_row(tab$row[i])) return(c(h1 = NA_real_, h2 = NA_real_))

  h0 <- c(tab$h01[i], tab$h02[i])
  h1 <- c(tab$h11[i], tab$h12[i])
  h0_ok <- all(is.finite(h0))
  h1_ok <- all(is.finite(h1))

  if (h0_ok && h1_ok && isTRUE(all.equal(h0, h1))) {
    return(c(h1 = h0[1], h2 = h0[2]))
  }
  if (h0_ok) return(c(h1 = h0[1], h2 = h0[2]))
  if (h1_ok) return(c(h1 = h1[1], h2 = h1[2]))

  c(h1 = NA_real_, h2 = NA_real_)
}

effective_sample_values <- function(tab, i) {
  if (is_aggregate_row(tab$row[i])) return(c(N.Co = NA_real_, N.Tr = NA_real_))
  c(N.Co = tab$N.Co[i], N.Tr = tab$N.Tr[i])
}

interval_limits <- function(tab) {
  # The output CSVs contain pointwise CIs and, when requested, uniform CBs.
  # For display, use CBs wherever available; WBATE/LBATE keep their CI limits.
  has_cb <- all(c("cb.lower", "cb.upper") %in% names(tab))
  use_cb <- rep(FALSE, nrow(tab))
  if (has_cb) {
    use_cb <- is.finite(tab$cb.lower) & is.finite(tab$cb.upper)
  }

  data.frame(
    lower = ifelse(use_cb, tab$cb.lower, tab$ci.lower),
    upper = ifelse(use_cb, tab$cb.upper, tab$ci.upper)
  )
}

################################## Output ######################################

display_rows <- function(tab, keep_rows = NULL) {
  aggregate_rows <- tab[is_aggregate_row(tab$row), , drop = FALSE]
  point_rows <- tab[!is_aggregate_row(tab$row), , drop = FALSE]

  if (!is.null(keep_rows)) {
    point_rows <- point_rows[point_rows$row %in% as.character(keep_rows), , drop = FALSE]
  }

  out <- rbind(point_rows, aggregate_rows)
  ints <- interval_limits(out)
  bws <- t(vapply(seq_len(nrow(out)), function(i) bandwidth_values(out, i), numeric(2)))
  ns <- t(vapply(seq_len(nrow(out)), function(i) effective_sample_values(out, i), numeric(2)))

  data.frame(
    label = vapply(out$row, fmt_row_label, character(1)),
    h1 = bws[, "h1"],
    h2 = bws[, "h2"],
    N.Co = ns[, "N.Co"],
    N.Tr = ns[, "N.Tr"],
    estimate = out$estimate.p,
    pvalue = out$p.value,
    lower = ints$lower,
    upper = ints$upper,
    stringsAsFactors = FALSE
  )
}

display_percent_rows <- function(tab) {
  aggregate_rows <- tab[is_aggregate_row(tab$row), , drop = FALSE]
  point_rows <- tab[!is_aggregate_row(tab$row), , drop = FALSE]
  out <- rbind(point_rows, aggregate_rows)
  bws <- t(vapply(seq_len(nrow(out)), function(i) bandwidth_values(out, i), numeric(2)))
  ns <- t(vapply(seq_len(nrow(out)), function(i) effective_sample_values(out, i), numeric(2)))

  data.frame(
    label = vapply(out$row, fmt_row_label, character(1)),
    h1 = bws[, "h1"],
    h2 = bws[, "h2"],
    N.Co = ns[, "N.Co"],
    N.Tr = ns[, "N.Tr"],
    itt = out$itt.percent,
    fuzzy = out$fuzzy.percent,
    stringsAsFactors = FALSE
  )
}

latex_body <- function(rows) {
  lines <- character(0)
  for (i in seq_len(nrow(rows))) {
    if (grepl("WBATE|LBATE", rows$label[i])) {
      lines <- c(lines, "  \\midrule")
    }
    lines <- c(lines, paste0(
      "   ", rows$label[i], " & ",
      fmt_num(rows$h1[i], 1), " & ",
      fmt_num(rows$h2[i], 1), " & ",
      fmt_num(rows$N.Co[i], 0), " & ",
      fmt_num(rows$N.Tr[i], 0), " & ",
      fmt_num(rows$estimate[i]), " & ",
      fmt_pvalue(rows$pvalue[i]), " & ",
      fmt_interval(rows$lower[i], rows$upper[i]), "\\\\"
    ))
  }
  lines
}

latex_percent_body <- function(rows) {
  lines <- character(0)
  for (i in seq_len(nrow(rows))) {
    if (grepl("WBATE|LBATE", rows$label[i])) {
      lines <- c(lines, "  \\midrule")
    }
    lines <- c(lines, paste0(
      "   ", rows$label[i], " & ",
      fmt_num(rows$h1[i], 1), " & ",
      fmt_num(rows$h2[i], 1), " & ",
      fmt_num(rows$N.Co[i], 0), " & ",
      fmt_num(rows$N.Tr[i], 0), " & ",
      fmt_num(rows$itt[i]), " & ",
      fmt_num(rows$fuzzy[i]), "\\\\"
    ))
  }
  lines
}

write_tabular <- function(rows, file) {
  lines <- c(
    "\\begin{tabular}{lccccccc}",
    "  \\toprule\\toprule",
    "   & $h_1$ & $h_2$ & $N_{\\mathrm{Co}}$ & $N_{\\mathrm{Tr}}$ & Estimate & $p$-value & 95\\% CI \\\\",
    "  \\midrule",
    latex_body(rows),
    "  \\bottomrule\\bottomrule",
    "\\end{tabular}"
  )
  writeLines(lines, file)
  invisible(lines)
}

write_percent_tabular <- function(rows, file) {
  lines <- c(
    "\\begin{tabular}{lcccccc}",
    "  \\toprule\\toprule",
    "   & $h_1$ & $h_2$ & $N_{\\mathrm{Co}}$ & $N_{\\mathrm{Tr}}$ & ITT / ITT.0 (\\%) & Fuzzy / ITT.0 (\\%) \\\\",
    "  \\midrule",
    latex_percent_body(rows),
    "  \\bottomrule\\bottomrule",
    "\\end{tabular}"
  )
  writeLines(lines, file)
  invisible(lines)
}

table_file <- function(id, suffix, small = FALSE) {
  paste0(id, if (isTRUE(small)) "_small" else "", "_", suffix, ".tex")
}

write_table_set <- function(id, kind, small, suffix) {
  tab <- read_result_table(id, kind, suffix)

  if (kind == "percent") {
    write_percent_tabular(display_percent_rows(tab), file.path(tables_dir, table_file(id, suffix)))
    return(invisible(TRUE))
  }

  write_tabular(
    display_rows(tab),
    file.path(tables_dir, table_file(id, suffix))
  )
  if (isTRUE(small)) {
    write_tabular(
      display_rows(tab, keep_rows = small_rows),
      file.path(tables_dir, table_file(id, suffix, small = TRUE))
    )
  }
}

################################ Sanity Checks #################################

check_generated_tables <- function(expected_tables) {
  missing_tables <- expected_tables[!file.exists(file.path(tables_dir, expected_tables))]
  if (length(missing_tables) > 0) {
    stop(sprintf("Missing expected tables: %s", paste(missing_tables, collapse = ", ")), call. = FALSE)
  }

  main_tables <- vapply(
    bw_specs$suffix,
    function(suffix) {
      paste(readLines(file.path(tables_dir, table_file("empapp_main_fuzzy", suffix))), collapse = "\n")
    },
    character(1)
  )
  main_table <- paste(main_tables, collapse = "\n")
  forbidden <- c("Z value", "t-statistic", "P>|t|", "P>|z|", "CI/CB", "$<$0.001")
  if (any(vapply(forbidden, grepl, logical(1), x = main_table, fixed = TRUE))) {
    stop("Generated empirical table contains an obsolete label.", call. = FALSE)
  }
  if (!grepl("WBATE", main_table) || !grepl("LBATE", main_table)) {
    stop("Generated empirical table is missing WBATE or LBATE.", call. = FALSE)
  }
  if (!grepl("$N_{\\mathrm{Co}}$", main_table, fixed = TRUE) || !grepl("$N_{\\mathrm{Tr}}$", main_table, fixed = TRUE)) {
    stop("Generated empirical table is missing N.Co or N.Tr headers.", call. = FALSE)
  }
}

################################### Run ########################################

clean_tables_dir()

for (j in seq_len(nrow(bw_specs))) {
  for (i in seq_len(nrow(table_specs))) {
    write_table_set(
      table_specs$id[i],
      table_specs$kind[i],
      table_specs$small[i],
      bw_specs$suffix[j]
    )
  }
}

expected_tables <- unlist(
  lapply(bw_specs$suffix, function(suffix) {
    c(
      vapply(table_specs$id, table_file, character(1), suffix = suffix),
      vapply(table_specs$id[table_specs$small], table_file, character(1), suffix = suffix, small = TRUE)
    )
  }),
  use.names = FALSE
)
check_generated_tables(expected_tables)

cat("Empirical application LaTeX tables complete.\n")
cat(sprintf("Tables written to: %s\n", tables_dir))

File: CTY_2026_JASA--empapp_tables.py
from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

SCRIPT_DIR = Path(__file__).resolve().parent
OUTPUT_DIR = SCRIPT_DIR / "output"
TABLES_DIR = SCRIPT_DIR / "tables"
SMALL_ROWS = {"1", "5", "10", "15", "20", "25", "30", "35", "40"}
AGGREGATE_LABELS = {"WBATE", "LBATE"}

TABLE_SPECS = [
    ("empapp_main_fuzzy", "effect", True),
    ("empapp_main_itt", "effect", True),
    ("empapp_main_fs", "effect", True),
    ("empapp_main_itt0", "effect", True),
    ("empapp_covbal_itt", "effect", True),
    ("empapp_main_percent", "percent", False),
]
BW_SPECS = [("main", "bwmain"), ("itt", "bwitt")]


def clean_tables_dir() -> None:
    TABLES_DIR.mkdir(parents=True, exist_ok=True)
    for old in TABLES_DIR.glob("empapp_*.tex"):
        old.unlink()


def read_result_table(stem: str, kind: str, suffix: str) -> pd.DataFrame:
    path = OUTPUT_DIR / f"{stem}_{suffix}.csv"
    if not path.exists():
        raise FileNotFoundError(f"Missing output file: {path}. Run CTY_2026_JASA--empapp.py first.")
    tab = pd.read_csv(path)
    required = ["row", "h01", "h02", "h11", "h12", "N.Co", "N.Tr"]
    if kind == "effect":
        required += ["estimate.p", "p.value", "ci.lower", "ci.upper"]
    elif kind == "percent":
        required += ["itt.percent", "fuzzy.percent"]
    else:
        raise ValueError(f"Unknown table kind: {kind}")
    missing = [name for name in required if name not in tab.columns]
    if missing:
        raise ValueError(f"{path} is missing columns: {', '.join(missing)}")
    tab["row"] = tab["row"].astype(str)
    return tab


def is_aggregate(row_label: str) -> bool:
    return row_label in AGGREGATE_LABELS


def scalar(value: object) -> float:
    try:
        out = float(value)
    except (TypeError, ValueError):
        return np.nan
    return out


def fmt_num(value: object, digits: int = 3) -> str:
    x = scalar(value)
    if not np.isfinite(x):
        return ""
    return f"{x:.{digits}f}"


def fmt_pvalue(value: object) -> str:
    x = scalar(value)
    if not np.isfinite(x):
        return ""
    if x < 0.001:
        return "0.000"
    return f"{x:.3f}"


def fmt_interval(lower: object, upper: object) -> str:
    lo = scalar(lower)
    hi = scalar(upper)
    if not np.isfinite(lo) or not np.isfinite(hi):
        return ""
    return f"({fmt_num(lo)}, {fmt_num(hi)})"


def fmt_row_label(row_id: str) -> str:
    if row_id == "WBATE":
        return "$\\mathtt{WBATE}$"
    if row_id == "LBATE":
        return "$\\mathtt{LBATE}$"
    return f"$\\bb_{{{row_id}}}$"


def bandwidth_values(row: pd.Series) -> tuple[float, float]:
    if is_aggregate(str(row["row"])):
        return np.nan, np.nan
    h0 = np.array([scalar(row["h01"]), scalar(row["h02"])])
    h1 = np.array([scalar(row["h11"]), scalar(row["h12"])])
    if np.all(np.isfinite(h0)) and np.all(np.isfinite(h1)) and np.allclose(h0, h1):
        return float(h0[0]), float(h0[1])
    if np.all(np.isfinite(h0)):
        return float(h0[0]), float(h0[1])
    if np.all(np.isfinite(h1)):
        return float(h1[0]), float(h1[1])
    return np.nan, np.nan


def effective_sample_values(row: pd.Series) -> tuple[float, float]:
    if is_aggregate(str(row["row"])):
        return np.nan, np.nan
    return scalar(row["N.Co"]), scalar(row["N.Tr"])


def interval_limits(tab: pd.DataFrame) -> tuple[pd.Series, pd.Series]:
    if {"cb.lower", "cb.upper"}.issubset(tab.columns):
        use_cb = np.isfinite(pd.to_numeric(tab["cb.lower"], errors="coerce")) & np.isfinite(
            pd.to_numeric(tab["cb.upper"], errors="coerce")
        )
        lower = tab["ci.lower"].copy()
        upper = tab["ci.upper"].copy()
        lower.loc[use_cb] = tab.loc[use_cb, "cb.lower"]
        upper.loc[use_cb] = tab.loc[use_cb, "cb.upper"]
        return lower, upper
    return tab["ci.lower"], tab["ci.upper"]


def order_rows(tab: pd.DataFrame, keep_rows: set[str] | None = None) -> pd.DataFrame:
    aggregates = tab.loc[tab["row"].isin(AGGREGATE_LABELS)].copy()
    points = tab.loc[~tab["row"].isin(AGGREGATE_LABELS)].copy()
    if keep_rows is not None:
        points = points.loc[points["row"].isin(keep_rows)].copy()
    return pd.concat([points, aggregates], ignore_index=True)


def display_rows(tab: pd.DataFrame, keep_rows: set[str] | None = None) -> pd.DataFrame:
    out = order_rows(tab, keep_rows=keep_rows)
    lower, upper = interval_limits(out)
    rows = []
    for i, row in out.iterrows():
        h1, h2 = bandwidth_values(row)
        nco, ntr = effective_sample_values(row)
        rows.append(
            {
                "label": fmt_row_label(str(row["row"])),
                "h1": h1,
                "h2": h2,
                "N.Co": nco,
                "N.Tr": ntr,
                "estimate": scalar(row["estimate.p"]),
                "pvalue": scalar(row["p.value"]),
                "lower": scalar(lower.iloc[i]),
                "upper": scalar(upper.iloc[i]),
            }
        )
    return pd.DataFrame(rows)


def display_percent_rows(tab: pd.DataFrame) -> pd.DataFrame:
    out = order_rows(tab)
    rows = []
    for _, row in out.iterrows():
        h1, h2 = bandwidth_values(row)
        nco, ntr = effective_sample_values(row)
        rows.append(
            {
                "label": fmt_row_label(str(row["row"])),
                "h1": h1,
                "h2": h2,
                "N.Co": nco,
                "N.Tr": ntr,
                "itt": scalar(row["itt.percent"]),
                "fuzzy": scalar(row["fuzzy.percent"]),
            }
        )
    return pd.DataFrame(rows)


def latex_body(rows: pd.DataFrame) -> list[str]:
    lines: list[str] = []
    for _, row in rows.iterrows():
        if "WBATE" in row["label"] or "LBATE" in row["label"]:
            lines.append("  \\midrule")
        lines.append(
            "   "
            + row["label"]
            + " & "
            + fmt_num(row["h1"], 1)
            + " & "
            + fmt_num(row["h2"], 1)
            + " & "
            + fmt_num(row["N.Co"], 0)
            + " & "
            + fmt_num(row["N.Tr"], 0)
            + " & "
            + fmt_num(row["estimate"])
            + " & "
            + fmt_pvalue(row["pvalue"])
            + " & "
            + fmt_interval(row["lower"], row["upper"])
            + "\\\\"
        )
    return lines


def latex_percent_body(rows: pd.DataFrame) -> list[str]:
    lines: list[str] = []
    for _, row in rows.iterrows():
        if "WBATE" in row["label"] or "LBATE" in row["label"]:
            lines.append("  \\midrule")
        lines.append(
            "   "
            + row["label"]
            + " & "
            + fmt_num(row["h1"], 1)
            + " & "
            + fmt_num(row["h2"], 1)
            + " & "
            + fmt_num(row["N.Co"], 0)
            + " & "
            + fmt_num(row["N.Tr"], 0)
            + " & "
            + fmt_num(row["itt"])
            + " & "
            + fmt_num(row["fuzzy"])
            + "\\\\"
        )
    return lines


def write_lines(path: Path, lines: list[str]) -> None:
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")


def write_tabular(rows: pd.DataFrame, path: Path) -> None:
    write_lines(
        path,
        [
            "\\begin{tabular}{lccccccc}",
            "  \\toprule\\toprule",
            "   & $h_1$ & $h_2$ & $N_{\\mathrm{Co}}$ & $N_{\\mathrm{Tr}}$ & Estimate & $p$-value & 95\\% CI \\\\",
            "  \\midrule",
            *latex_body(rows),
            "  \\bottomrule\\bottomrule",
            "\\end{tabular}",
        ],
    )


def write_percent_tabular(rows: pd.DataFrame, path: Path) -> None:
    write_lines(
        path,
        [
            "\\begin{tabular}{lcccccc}",
            "  \\toprule\\toprule",
            "   & $h_1$ & $h_2$ & $N_{\\mathrm{Co}}$ & $N_{\\mathrm{Tr}}$ & ITT / ITT.0 (\\%) & Fuzzy / ITT.0 (\\%) \\\\",
            "  \\midrule",
            *latex_percent_body(rows),
            "  \\bottomrule\\bottomrule",
            "\\end{tabular}",
        ],
    )


def table_file(stem: str, suffix: str, small: bool = False) -> str:
    return f"{stem}{'_small' if small else ''}_{suffix}.tex"


def write_table_set(stem: str, kind: str, small: bool, suffix: str) -> None:
    tab = read_result_table(stem, kind, suffix)
    if kind == "percent":
        write_percent_tabular(display_percent_rows(tab), TABLES_DIR / table_file(stem, suffix))
        return
    write_tabular(display_rows(tab), TABLES_DIR / table_file(stem, suffix))
    if small:
        write_tabular(display_rows(tab, keep_rows=SMALL_ROWS), TABLES_DIR / table_file(stem, suffix, small=True))


def check_generated_tables(expected: list[str]) -> None:
    missing = [name for name in expected if not (TABLES_DIR / name).exists()]
    if missing:
        raise FileNotFoundError(f"Missing expected tables: {', '.join(missing)}")
    main_text = "\n".join((TABLES_DIR / table_file("empapp_main_fuzzy", suffix)).read_text(encoding="utf-8") for _, suffix in BW_SPECS)
    forbidden = ["Z value", "t-statistic", "P>|t|", "P>|z|", "CI/CB", "$<$0.001"]
    if any(label in main_text for label in forbidden):
        raise ValueError("Generated empirical table contains an obsolete label.")
    if "WBATE" not in main_text or "LBATE" not in main_text:
        raise ValueError("Generated empirical table is missing WBATE or LBATE.")
    if "$N_{\\mathrm{Co}}$" not in main_text or "$N_{\\mathrm{Tr}}$" not in main_text:
        raise ValueError("Generated empirical table is missing N.Co or N.Tr headers.")


def main() -> None:
    clean_tables_dir()
    for _, suffix in BW_SPECS:
        for stem, kind, small in TABLE_SPECS:
            write_table_set(stem, kind, small, suffix)
    expected: list[str] = []
    for _, suffix in BW_SPECS:
        expected.extend(table_file(stem, suffix) for stem, _, _ in TABLE_SPECS)
        expected.extend(table_file(stem, suffix, small=True) for stem, _, small in TABLE_SPECS if small)
    check_generated_tables(expected)
    print("Empirical application LaTeX tables complete.")
    print(f"Tables written to: {TABLES_DIR.relative_to(SCRIPT_DIR)}")


if __name__ == "__main__":
    main()
